weakmap.c

```diff
@@ -414,6 +422,60 @@ nonspecial_obj_id(VALUE obj)
 #endif
 }
 
+struct wmap_aset_replace_args {
+    VALUE new_value;
+    VALUE old_value;
+};
+
+static int
+wmap_aset_replace_value(st_data_t *key, st_data_t *val, st_data_t _args, int existing)
+{
+    struct wmap_aset_replace_args *args = (struct wmap_aset_replace_args *)_args;
+
+    if (existing) {
+        args->old_value = *val;
+    }
+    *val = (st_data_t)args->new_value;
+    return ST_CONTINUE;
+}
+
+static int
+wmap_aset_remove_inverse_ref(st_data_t *key, st_data_t *val, st_data_t arg, int existing)
```

Comment:
So I merged both together, and I realized there was a minor difference.

Since `wmap_final_func` is called from finalizers, ordering isn't exactly stable so the backref migth already be done.

`wmap_aset_remove_inverse_ref` though had some asserts to ensure the ref was found and removed.

In the end I don't think it was that important, so I merged both anyway, without the asserts.